hpe3parclient/client.py

```diff
@@ -508,46 +508,65 @@ def createVolume(self, name, cpgName, sizeMiB, optional=None):
+        # replacing compression+tdvv with reduce key for DECO
         if not optional and self.primera_supported:
-            optional = {}
-            optional['tpvv'] = True
+            optional = {'tpvv': True}
         if optional:
             if self.primera_supported:
-                if 'tdvv' in optional:
-                    optional.pop('tdvv')
-
-                if optional.get('tpvv') is True \
+                for key in ['tpvv', 'compression', 'tdvv']:
+                    option = optional.get(key)
+                    if option and option not in [True, False]:
+                        # raising exception for junk compression input
+                        ex_desc = "39 - invalid input: wrong type for value"
+                        raise exceptions.HTTPBadRequest(ex_desc)
+
+                combination = ['tdvv', 'compression']
+                len_diff = len(set(combination) - set(optional.keys()))
+                if len_diff == 1:
+                    raise exceptions.HTTPBadRequest("Invalid input:For \
+                        Deco volumes, both 'tdvv' and 'compression' \
+                        must be specified")
+                if optional.get('tdvv') is True \
                         and optional.get('compression') is True:
                     optional['reduce'] = True
 
-                if not optional.get('tpvv') \
-                        and not optional.get('compression'):
-                    optional['tpvv'] = True
+                if optional.get('tdvv') is False \
+                        and optional.get('compression') is False:
+                    optional['reduce'] = False
 
-                if 'compression' in optional:
-                    if optional.get('compression') is not None:
-                        if optional.get('compression') is True:
-                            optional['reduce'] = True
-                        elif optional.get('compression') is False:
-                            optional['reduce'] = False
-                        else:
-                            # raising exception for junk compression input
-                            ex_desc = "39 - invalid input: wrong type for value \
-                                       - compression"
-                            raise exceptions.HTTPBadRequest(ex_desc)
+                if optional.get('tdvv') is False \
+                        and optional.get('compression') is True:
+                    raise exceptions.HTTPBadRequest("For enabling the \
+                        compression for primera tdvv and compression \
+                        both flags must be true")
 
-                    optional.pop('compression')
+                if optional.get('tdvv') is True \
+                        and optional.get('compression') is False:
+                    raise exceptions.HTTPBadRequest("For enabling the \
+                        compression for primera tdvv and compression \
+                        both are true")
```

Comment:
For consistency, let's keep the same message as above:

**For Deco volumes, both 'tdvv' and 'compression' must be specified**